### ISSUE TITLE
Issue with CheckTx propogation using ABCI++ socket interface

### ISSUE THREAD

**Issue by ethanfrey** [cometbft/cometbft#4094]
### Setup

**CometBFT version** (use `cometbft version` or `git rev-parse --verify HEAD` if installed from source):

v0.38.12 (via docker build)

**Have you tried the latest version**: yes/no

yes

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

layer-sdk - Rust application, not publically available.
Connects using tendermint-abci sockets

**Environment**:
- **OS** (e.g. from /etc/os-release): Linux / Ubuntu 22.04
- **Install tools**:
- **Others**:

**node command runtime flags**:

`cometbft start`

### What happened?

I connect a local Rust ABCI node to CometBFT (both as local docker containers, via docker compose).
On a single node, "everything" works fine. Well, except after a check_tx returns a success, the tx is neither added to mempool (not present in `prepare_proposal`) nor propogated when I connect to a peer.

Early on, during some 0.38.0-rc.X, I saw this, and cached the approved tx locally and fed it in `prepare_proposal`. The app runs "fine" like this, but only single node.

Recently, I attempted to connect a second non-signing node via p2p, and it is able to run all blocks. However, when I submit a tx to the rpc of the non-validating full-node, it never appears in the mempool of the validating node.

I return the following struct back to cometbft in response to checktx: 

`ResponseCheckTx { code: 0, data: b"", log: "[]", info: "", gas_wanted: 2270289, gas_used: 2270289, events: [], codespace: "sdk" }`

Note: I just added codespace (was empty before) to try to debug.

### What did you expect to happen?

I expected it to enter the mempool properly

### How to reproduce it

I cannot (yet) share how to reproduce, but happy to get some info so I can provide a better report.

First, I was wondering if you have seen a similar error before or have any input on the return values. Is there something I am obviously missing and should do differently?

Second, I am happy to turn up specific logging if there is something else I can provide you (let me know what to do and I will attach). There is no mention of check_tx on either node with info level logging on cometbft.

Third, I was wondering if you had any minimal tests using abci socket to external process I could see as a test, and then try to adapt it into a test case for my issue.

**Comment by ethanfrey** [comment]
Hey @melekes i know this is impossible to debug on your side with this info. Could you provide me any logs/debugging info I could look at cometbft side to understand the mempool/checktx interactions better (which log flags to enable, not all as it is too verbose to see anything)

I would like to produce a minimal test case to recreate this. Ideally with Rust over abci socket. Do you have some minimal dummy app in rust I could use as a starting point?

**Comment by ethanfrey** [comment]
I would like to make a minimal test case with https://github.com/informalsystems/tendermint-rs/blob/main/abci/src/application/kvstore/main.rs

and comet bft as two different processes, so I can see if the bug is on “your side” or “my side”. And if it is in my code, at least have a minimal example that works, which I can start to debug from

I dug through the repo looking for some test running this over abci sockets, but I cannot find it anywhere. Is there any e2e test for the rust abci server implementation?

**Comment by cason** [comment]
You should try using the debug level logging on the mempool. But unfortunately, a lot of debug log entries were removed from the mempool for performance purposes (I think), so you'll probably need to manually add some.

You can also try the [/unconfirmed_txs](https://docs.cometbft.com/v0.38/rpc/#/Info/unconfirmed_txs) RPC endpoint.

**Comment by ethanfrey** [comment]
After playing around with this for a while, it turns out that the issue was due to block max_gas parameter being set to -1, which was the default in `comet init` (at least in May). Now, that is treated as non-gas rather than infinite I think, so the tx were all being dropped.

I never did find logs for that, but after trying to debug comparing to sample apps, we set max_gas to 10000000, and it suddenly worked.

Not sure when this behavior changed, but it would be nice to document it better (ideally error on negative values parsing the genesis if they change behavior)

**Comment by cason** [comment]
We have it documented here: https://docs.cometbft.com/v0.38/spec/abci/abci++_app_requirements#blockparamsmaxgas

> BlockParams.MaxGas
> 
> The maximum of the sum of GasWanted that will be allowed in a proposed block. This is not enforced by the consensus algorithm. It is left to the Application to enforce (ie. if transactions are included past the limit, they should return non-zero codes). It is used by CometBFT to limit the transactions included in a proposed block.
>
> Must have MaxGas >= -1. If MaxGas == -1, no limit is enforced.

**Comment by cason** [comment]
But you are saying that `-1` has changed behavior? If so we definitely have to check and fix.

**Comment by ethanfrey** [comment]
I know those docs and that behavior (for context, I have used tendermint since 2017 and used to work at AiB).

I also know -1 is only good for local testing / devnets, and anything public needs a limit or you can create fun DoS attacks on the chain.

However, my last test was showing me that a 90000 gas tx was consistently dropped from the mempool when it passed check_tx when mas_gas was -1. And when I changed it to 10000000, it suddenly worked (same binary).

If you could tell me what log flags to enable, I could show you the cometbft logs.
But I bet you could make a testcase for this

**Comment by cason** [comment]
> I also know -1 is only good for local testing / devnets, and anything public needs a limit or you can create fun DoS attacks on the chain.

Absolutely. The point is how gas is used by different networks, that has never clear to me (I'd say, us).

**Comment by cason** [comment]
> If you could tell me what log flags to enable, I could show you the cometbft logs.

As I said, unfortunately a lot of refactoring in mempool code has removed most of the debug log entries, because they were costly. We are fixing that in recent issues/PRs. So with upstream code I am afraid that the only way to debug is by relying on the response returned to the RPC client.

**Comment by cason** [comment]
>  But I bet you could make a testcase for this

Would you mind to open an issue describing this situation in more detail. We probably can reproduce the scenario, if it is indeed valid, and find solutions.
